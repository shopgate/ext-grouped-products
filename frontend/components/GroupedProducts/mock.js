import set from 'lodash/set';
import get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';

const availability = {
  text: 'Lorem Ipsum',
  state: 'ok',
};

const stock = {
  ignoreQuantity: false,
  quantity: 30,
  info: 'Info',
  orderable: true,
  minOrderQuantity: 1,
  maxOrderQuantity: 1000,
};

const price = {
  currency: 'EUR',
  unitPrice: 5.0,
  unitPriceMin: 0,
  unitPriceStriked: 0,
  msrp: 0,
  discount: 0,
};

const product = {
  id: '1337',
  name: 'A shiny product',
  featuredImageUrl: 'https://cdn.acme.com/kitten.jpg',
  availability,
  price,
  stock,
};

let discounted = cloneDeep({
  ...product,
  id: '4711',
  name: 'A discounted product',
});

discounted = set(discounted, ['price', 'msrp'], 10.0);
discounted = set(discounted, ['price', 'discount'], 50);

let notOrderable = cloneDeep({
  ...product,
  id: '1234',
  name: 'A not orderable product',
});

notOrderable = set(notOrderable, ['stock', 'orderable'], false);
notOrderable = set(notOrderable, ['stock', 'quantity'], 0);

const mockedState = {
  product: {
    currentProduct: {
      productId: '1337',
    },
    productsById: {
      1337: {
        productData: product,
      },
    },
  },
};

export {
  mockedState,
  product as mockedProduct,
  discounted as mockedDiscountedProduct,
  notOrderable as mockedNotOrderableProduct,
};
