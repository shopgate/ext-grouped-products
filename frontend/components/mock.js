import set from 'lodash/set';
import cloneDeep from 'lodash/cloneDeep';
import { generateHash } from '../helpers';

const availability = {
  text: 'Lorem Ipsum',
  state: 'ok',
};

const stock = {
  ignoreQuantity: false,
  quantity: 4,
  info: 'Info',
  orderable: true,
  minOrderQuantity: 1,
  maxOrderQuantity: 1000,
};

const price = {
  currency: 'EUR',
  unitPrice: 5.0,
  unitPriceMin: 0,
  unitPriceStriked: 0,
  msrp: 0,
  discount: 0,
  info: '',
};

const product = {
  id: '1337',
  name: 'A shiny product',
  featuredImageUrl: 'https://cdn.acme.com/kitten.jpg',
  flags: {
    hasChildren: true,
    hasVariants: false,
  },
  availability,
  price,
  stock,
};

// eslint-disable-next-line import/no-mutable-exports
let discountedMsrp = cloneDeep({
  ...product,
  id: '4711',
  name: 'A discounted product with msrp',
});

discountedMsrp = set(discountedMsrp, ['price', 'msrp'], 10.0);
discountedMsrp = set(discountedMsrp, ['price', 'discount'], 50);

// eslint-disable-next-line import/no-mutable-exports
let discountedStrike = cloneDeep({
  ...product,
  id: '1985',
  name: 'A discounted product with stike price and info',
});

discountedStrike = set(discountedStrike, ['price', 'unitPriceStriked'], 10.0);
discountedStrike = set(discountedStrike, ['price', 'discount'], 50);
discountedStrike = set(discountedStrike, ['price', 'info'], '5 EUR / kg');

// eslint-disable-next-line import/no-mutable-exports
let notOrderable = cloneDeep({
  ...product,
  id: 'foo',
  name: 'A not orderable product',
});

notOrderable = set(notOrderable, ['stock', 'orderable'], false);
notOrderable = set(notOrderable, ['stock', 'quantity'], 0);

// eslint-disable-next-line import/no-mutable-exports
let ignoredQuantity = cloneDeep({
  ...product,
  id: '1234',
  name: 'A product with ignored quantity',
});

ignoredQuantity = set(ignoredQuantity, ['stock', 'ignoreQuantity'], true);
ignoredQuantity = set(ignoredQuantity, ['stock', 'quantity'], 30);

const mockedState = {
  product: {
    currentProduct: {
      productId: '1337',
    },
    resultsByHash: {
      [generateHash('1337')]: {
        isFetching: false,
        products: [
          product.id,
          discountedMsrp.id,
          discountedStrike.id,
          notOrderable.id,
          ignoredQuantity.id,
        ],
      },
    },
    productsById: {
      [product.id]: {
        productData: product,
      },
      [discountedMsrp.id]: {
        productData: discountedMsrp,
      },
      [discountedStrike.id]: {
        productData: discountedStrike,
      },
      [notOrderable.id]: {
        productData: notOrderable,
      },
      [ignoredQuantity.id]: {
        productData: ignoredQuantity,
      },
    },
  },
  favorites: {
    products: {
      isFetching: false,
      ids: ['1337'],
    },
  },
};

const mockedProps = {
  handleAddToCart: () => { },
  isFavorite: false,
  addToCartButtonProps: {},
  isAddToCartButtonVisible: false,
  product: {},
  onClose: () => { }
};

const mockedStateWithRegularProduct = set(
  cloneDeep(mockedState),
  ['product', 'productsById', product.id, 'productData', 'flags'],
  {
    hasChildren: false,
    hasVariants: false,
  }
);

const mockedPropsWithRegularProduct = {
  handleAddToCart: () => { },
  isFavorite: true,
  addToCartButtonProps: {},
  isAddToCartButtonVisible: true,
  product: {},
};

const mockedStateWithoutCurrentProductData = set(
  cloneDeep(mockedState),
  ['product', 'productsById', product.id],
  undefined
);

const mockedPropsWithoutCurrentProduct = {
  handleAddToCart: () => { },
  isFavorite: true,
  addToCartButtonProps: {},
  isAddToCartButtonVisible: true,
  product: {},
};

export {
  mockedProps,
  mockedPropsWithRegularProduct,
  mockedPropsWithoutCurrentProduct,
  mockedState,
  mockedStateWithRegularProduct,
  mockedStateWithoutCurrentProductData,
  product as mockedProduct,
  discountedMsrp as mockedMsrpProduct,
  discountedStrike as mockedStrikePriceProduct,
  notOrderable as mockedNotOrderableProduct,
  ignoredQuantity as mockedIgnoredQuantityProduct,
};
