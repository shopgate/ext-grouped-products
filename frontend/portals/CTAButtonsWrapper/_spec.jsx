import React from 'react';
import { shallow } from 'enzyme';
import CTAButtons from './index';

describe('<CTAButtonsWrapper />', () => {
  it('should render', () => {
    const wrapper = shallow(<CTAButtons />);
    expect(wrapper).toMatchSnapshot();
  });
});
