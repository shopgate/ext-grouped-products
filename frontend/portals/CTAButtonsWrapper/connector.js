import { connect } from 'react-redux';
import {
  getCurrentProduct,
  isProductOrderable,
} from '@shopgate/pwa-common-commerce/product/selectors/product';
import { isCurrentProductOnFavoriteList } from '@shopgate/pwa-common-commerce/favorites/selectors';
import {
  isProductPageLoading,
  isProductPageOrderable,
} from '@shopgate/pwa-common-commerce/product/selectors/page';
import { addProductToCart } from './actions';
import { isMainAddToCartButtonVisible } from '../../selectors/index';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @param {Object} props The current props.
 * @return {Object} The extended component props.
 */
const mapStateToProps = (state, props) => ({
  addToCartButtonProps: {
    isLoading: isProductPageLoading(state, props),
    isOrderable: isProductPageOrderable(state, props),
    isDisabled: !isProductOrderable(state, props),
  },
  isAddToCartButtonVisible: isMainAddToCartButtonVisible(state, props),
  isFavorite: isCurrentProductOnFavoriteList(state, props),
  product: getCurrentProduct(state, props) || {},
});

/**
 * Connects the dispatch function to a callable function in the props.
 * @param {Function} dispatch The redux dispatch function.
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  handleAddToCart: data => dispatch(addProductToCart(data)),
});

export default connect(mapStateToProps, mapDispatchToProps);
