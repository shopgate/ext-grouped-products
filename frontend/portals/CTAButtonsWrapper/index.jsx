import React from 'react';
import { RouteContext } from '@shopgate/pwa-common/context';
import { hex2bin } from '@shopgate/pwa-common/helpers/data';
import connect from './connector';

import CTAButtons from '../../components/CTAButtons';

export const CTAButtonsContainer = connect(({
  addToCartButtonProps,
  isAddToCartButtonVisible,
  isFavorite,
  product,
  handleAddToCart,
}) => (
  <CTAButtons
    addToCartButtonProps={addToCartButtonProps}
    isAddToCartButtonVisible={isAddToCartButtonVisible}
    isFavorite={isFavorite}
    product={product}
    handleAddToCart={handleAddToCart}
  />
));

/**
 * Renders CTAButtons component in product.ctas portal
 * @param {Object} params RouteContext params
 * @returns {JSX}
 */
const CTAButtonsWrapper = () => (
  <RouteContext.Consumer>
    {({ params }) => (
      <div>
        <CTAButtonsContainer productId={hex2bin(params.productId)} />
      </div>
    )}
  </RouteContext.Consumer>
);

export default CTAButtonsWrapper;
