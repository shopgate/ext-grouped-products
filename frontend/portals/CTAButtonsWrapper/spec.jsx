import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import mockRenderOptions from '@shopgate/pwa-common/helpers/mocks/mockRenderOptions';
import { basicProductState } from '@shopgate/pwa-common-commerce/product/mock';
import { ThemeContext } from '@shopgate/pwa-common/context';
import CTAButtonsWrapper from './index';

const mockedStore = configureStore();
jest.mock('../../selectors/index', () => ({
  isMainAddToCartButtonVisible: jest.fn().mockReturnValue(true),
}));
jest.mock('../../components/CTAButtons', () => () => (<div>CTAButtons</div>));
const ProductContext = React.createContext({
  options: {},
  productId: 'mock',
  variantId: null,
  conditioner: {},
});
const mockThemeContext = {
  contexts: {
    ProductContext,
  },
};
/**
 * Creates Component
 * @return {ReactWrapper}
 */
const createComponent = () => mount(
  <ThemeContext.Provider value={mockThemeContext}>
    <Provider store={mockedStore({ product: basicProductState })}>
      <CTAButtonsWrapper />
    </Provider>
  </ThemeContext.Provider>,
  mockRenderOptions
);

describe('CTAButtons', () => {
  it('should render', () => {
    const component = createComponent();
    expect(component.find('div').exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});
