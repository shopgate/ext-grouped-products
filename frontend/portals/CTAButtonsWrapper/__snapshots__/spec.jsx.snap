// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CTAButtonsWrapper /> should render 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
